crates/gitbutler-core/src/virtual_branches/branch/hunk.rs

```diff
@@ -157,12 +154,35 @@ impl Hunk {
         self.start == other.new_start && self.end == other.new_start + other.new_lines
     }
 
-    pub fn hash(diff: &str) -> String {
-        let addition = diff
-            .lines()
-            .skip(1) // skip the first line which is the diff header
-            .collect::<Vec<_>>()
-            .join("\n");
-        format!("{:x}", md5::compute(addition))
+    /// Produce a hash from `diff` as hex-string, which is **assumed to have a one-line diff header**!
+    ///
+    /// ### Panics
+    /// It skips the first line which it assumes to tbe the diff-header.
+    /// However, if there is only one line, this function panics.
+    ///
+    /// ### Notes on Persistence
+    /// Note that there is danger in changing the hash function as this information is persisted
+    /// in the virtual-branch toml file. Even if it can still be parsed or decoded,
+    /// these values have to remain consistent.
+    pub fn hash_diff(diff: &BStr) -> HunkHash {
+        let mut ctx = md5::Context::new();
+        let mut hashed_something = false;
+        diff.lines()
+            .skip(1) // skip the first line which is the diff header.
+            .for_each(|line| {
+                hashed_something = true;
+                ctx.consume(line)
+            });
+        assert!(
+            hashed_something,
+            "BUG: input must have a first line, but nothing was hashed here (len={}): {diff}",
```

Comment:
When putting it to the test, I noticed that the hash function can be called with empty `diff` input, something that shouldn't be possible.
After all, this hash function has requirements to its input and only then will it produce the expected hash.
Now I am not sure at all if this is called consistently, and since assertion is hit during tests something about these hashes seems fishy. The question now is if they are used consistently enough for the system to work. Maybe related to #3451 .

Reply:
And what makes it worse is that these hashes are used quite interchangeably: like in [`Hunk::from(GitHunk)`](https://github.com/gitbutlerapp/gitbutler/blob/7c2e25a9e4c63581144faf403ebd0424109831af/crates/gitbutler-core/src/virtual_branches/branch/hunk.rs#L23).

Reply:
This was my mistake - it's probably OK to have empty hunks, as long as they indeed have a `@@` header line.